src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2998,167 +3303,75 @@ public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) {
             return futures;
         } catch (Throwable t) {
             for (Future<T> e : futures)
-                ForkJoinTask.cancelIgnoringExceptions(e);
+                e.cancel(true);
             throw t;
         }
     }
 
-    @Override
-    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks,
-                                         long timeout, TimeUnit unit)
+    /**
+     * Common support for timed and untimed invokeAll
+     */
+    private  <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks,
+                                           long deadline)
         throws InterruptedException {
-        long nanos = unit.toNanos(timeout);
         ArrayList<Future<T>> futures = new ArrayList<>(tasks.size());
         try {
             for (Callable<T> t : tasks) {
-                ForkJoinTask<T> f =
-                    new ForkJoinTask.AdaptedInterruptibleCallable<T>(t);
+                ForkJoinTask<T> f = ForkJoinTask.adaptInterruptible(t);
                 futures.add(f);
                 poolSubmit(true, f);
             }
-            long startTime = System.nanoTime(), ns = nanos;
-            boolean timedOut = (ns < 0L);
-            for (int i = futures.size() - 1; i >= 0; --i) {
-                ForkJoinTask<T> f = (ForkJoinTask<T>)futures.get(i);
-                if (!f.isDone()) {
-                    if (!timedOut)
-                        timedOut = !f.quietlyJoin(ns, TimeUnit.NANOSECONDS);
-                    if (timedOut)
-                        ForkJoinTask.cancelIgnoringExceptions(f);
-                    else
-                        ns = nanos - (System.nanoTime() - startTime);
-                }
-            }
+            for (int i = futures.size() - 1; i >= 0; --i)
```

Comment:
@DougLea Might not be necessary for performance, but using an ArrayDeque and using `pollLast()` might be cleaner?

Reply:
This was a good enough idea that I started doing it until realizing/remembering that ArrayDeque doesn't implement List so can't be used here (without making a copy). Oh well.

Reply:
@DougLea Ouch—I completely missed that it is also returned. Apologies :(